File: ios/haishin_kit/Sources/StreamSessionHandler.swift
import Foundation
#if canImport(Flutter)
import Flutter
#endif
#if canImport(FlutterMacOS)
import FlutterMacOS
#endif
import HaishinKit
import RTMPHaishinKit
import AVFoundation
import VideoToolbox

final class StreamSessionHandler: NSObject {
    enum ErrorCode: String {
        case connectFailed  = "CONNECT_FAILED"
    }

    private let plugin: HaishinKitPlugin
    private var session: Session?
    private var texture: StreamFlutterTexture?
    private var channel: FlutterEventChannel?
    private var eventSink: FlutterEventSink?
    private var subscription: Task<(), Error>? {
        didSet {
            oldValue?.cancel()
        }
    }

    init(plugin: HaishinKitPlugin, session: Session) {
        self.plugin = plugin
        self.session = session
        super.init()
        let id = Int(bitPattern: ObjectIdentifier(self))
        if let messanger = plugin.registrar?.messenger() {
            self.channel = FlutterEventChannel(name: "com.haishinkit.eventchannel/\(id)", binaryMessenger: messanger)
        } else {
            self.channel = nil
        }
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            self.channel?.setStreamHandler(self)
        }
        subscription = Task { [weak self] in
            for await status in await session.readyState {
                DispatchQueue.main.async { [eventSink = self?.eventSink] in
                    eventSink?("\(status)")
                }
            }
        }
    }
}

extension StreamSessionHandler: MethodCallHandler {
    // MARK: MethodCallHandler
    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard
            let arguments = call.arguments as? [String: Any?] else {
            result(nil)
            return
        }
        switch call.method {
        case "StreamSession#setAudioSettings":
            guard
                let settings = arguments["settings"] as? [String: Any?] else {
                result(nil)
                return
            }
            if let bitrate = settings["bitrate"] as? NSNumber {
                Task {
                    var audioSettings = await session?.stream.audioSettings ?? .default
                    audioSettings.bitRate = bitrate.intValue
                    _ = try? await session?.stream.setAudioSettings(audioSettings)
                    result(nil)
                }
                return
            }
            result(nil)
        case "StreamSession#setVideoSettings":
            guard
                let settings = arguments["settings"] as? [String: Any?] else {
                result(nil)
                return
            }
            Task {
                var videoSettings = await session?.stream.videoSettings ?? .default
                if let bitrate = settings["bitrate"] as? NSNumber {
                    videoSettings.bitRate = bitrate.intValue
                }
                if let width = settings["width"] as? NSNumber, let height = settings["height"] as? NSNumber {
                    videoSettings.videoSize = CGSize(width: .init(width.intValue), height: .init(height.intValue))
                }
                if let frameInterval = settings["frameInterval"] as? NSNumber {
                    videoSettings.maxKeyFrameIntervalDuration = frameInterval.int32Value
                }
                if let profileLevel = settings["profileLevel"] as? String {
                    videoSettings.profileLevel = ProfileLevel(rawValue: profileLevel)?.kVTProfileLevel ?? ProfileLevel.H264_Baseline_AutoLevel.kVTProfileLevel
                }
                _ = try? await session?.stream.setVideoSettings(videoSettings)
                result(nil)
            }
        case "StreamSession#registerTexture":
            guard
                let registry = plugin.registrar?.textures() else {
                result(nil)
                return
            }
            if let texture {
                result(texture.id)
            } else {
                let texture = StreamFlutterTexture(registry: registry)
                self.texture = texture
                Task {
                    await session?.stream.addOutput(texture)
                    result(texture.id)
                }
            }
        case "StreamSession#unregisterTexture":
            guard
                let registry = plugin.registrar?.textures() else {
                result(nil)
                return
            }
            if let textureId = arguments["id"] as? Int64 {
                registry.unregisterTexture(textureId)
            }
            result(nil)
        case "StreamSession#updateTextureSize":
            guard let _ = plugin.registrar?.textures() else {
                result(nil)
                return
            }
            if let texture {
                if let width = arguments["width"] as? NSNumber,
                   let height = arguments["height"] as? NSNumber {
                    texture.bounds = CGSize(width: width.doubleValue, height: height.doubleValue)
                }
                result(texture.id)
            } else {
                result(nil)
            }
        case "StreamSession#connect":
            Task {
                do {
                    try await session?.connect {
                    }
                    result(nil)
                } catch {
                    result(FlutterError(code: ErrorCode.connectFailed.rawValue, message: nil, details: nil))
                }
            }
        case "StreamSession#close":
            Task {
                _ = try? await session?.close()
                result(nil)
            }
        case "StreamSession#dispose":
            subscription = nil
            texture = nil
            session = nil
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }
}

extension StreamSessionHandler: FlutterStreamHandler {
    // MARK: FlutterStreamHandler
    public func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    public func onCancel(withArguments arguments: Any?) -> FlutterError? {
        return nil
    }
}

File: android/src/main/kotlin/com/haishinkit/haishin_kit/StreamSessionHandler.kt
package com.haishinkit.haishin_kit

import android.media.MediaFormat.KEY_LEVEL
import android.media.MediaFormat.KEY_PROFILE
import android.os.Build
import android.util.Log
import android.util.Size
import com.haishinkit.codec.CodecOption
import com.haishinkit.stream.StreamSession
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

class StreamSessionHandler(
    private val plugin: HaishinKitPlugin,
    var session: StreamSession?,
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    companion object Companion {
        private const val TAG = "StreamSession"
        const val CONNECT_FAILED = "CONNECT_FAILED"
    }

    private var channel = EventChannel(
        plugin.flutterPluginBinding.binaryMessenger, "com.haishinkit.eventchannel/${hashCode()}"
    )
    private var eventSink: EventChannel.EventSink? = null
        set(value) {
            field?.endOfStream()
            field = value
        }
    private var texture: StreamViewTexture? = null
        set(value) {
            field?.let {
                session?.stream?.unregisterOutput(it)
            }
            field?.dispose()
            field = value
        }
    private var scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)

    init {
        channel.setStreamHandler(this)
        session?.readyState?.onEach { state ->
            eventSink?.success(state.toString().lowercase())
        }?.launchIn(scope)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        Log.d(TAG, "onMethodCall: " + call.method + "()")
        if (session == null) {
            result.error(HaishinKitPlugin.ILLEGAL_STATE, "session is disposed.", null)
            return
        }
        when (call.method) {
            "$TAG#setFrameRate" -> {
                val value = call.argument<Int?>("value")
                if (value == null) {
                    result.error(HaishinKitPlugin.INVALID_ARGUMENTS, null, null)
                } else {
                    session?.stream?.videoSetting?.frameRate = value
                    result.success(null)
                }
            }

            "$TAG#setAudioSettings" -> {
                val source = call.argument<Map<String, Any?>>("settings")
                if (source == null) {
                    result.error(HaishinKitPlugin.INVALID_ARGUMENTS, null, null)
                } else {
                    (source["bitrate"] as? Int)?.let {
                        session?.stream?.audioSetting?.bitRate = it
                    }
                    result.success(null)
                }
            }

            "$TAG#setVideoSettings" -> {
                val source = call.argument<Map<String, Any?>>("settings") ?: return
                (source["width"] as? Int)?.let {
                    session?.stream?.videoSetting?.width = it
                }
                (source["height"] as? Int)?.let {
                    session?.stream?.videoSetting?.height = it
                }
                (source["frameInterval"] as? Int)?.let {
                    session?.stream?.videoSetting?.IFrameInterval = it
                }
                (source["bitrate"] as? Int)?.let {
                    session?.stream?.videoSetting?.bitRate = it
                }
                (source["profileLevel"] as? String)?.let {
                    try {
                        val profileLevel = ProfileLevel.valueOf(it)
                        val options = mutableListOf<CodecOption>()
                        options.add(CodecOption(KEY_PROFILE, profileLevel.profile))
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                            options.add(CodecOption(KEY_LEVEL, profileLevel.level))
                        }
                        session?.stream?.videoSetting?.options = options
                    } catch (ignored: Exception) {
                        // Do nothing, use default setting
                    }
                }
                result.success(null)
            }

            "$TAG#registerTexture" -> {
                val texture = StreamViewTexture(plugin.flutterPluginBinding)
                this.texture = texture
                session?.stream?.registerOutput(texture)
                result.success(texture.id)
            }

            "$TAG#unregisterTexture" -> {
                texture = null
                result.success(null)
            }

            "$TAG#updateTextureSize" -> {
                val width = call.argument<Double>("width") ?: 0
                val height = call.argument<Double>("height") ?: 0
                texture?.imageExtent = Size(width.toInt(), height.toInt())
                result.success(texture?.id)
            }

            "$TAG#connect" -> {
                scope.launch {
                    session?.connect()?.onSuccess {
                        result.success(null)
                    }?.onFailure {
                        result.error(CONNECT_FAILED, null, null)
                    }
                }
            }

            "$TAG#close" -> {
                scope.launch {
                    session?.close()
                    result.success(null)
                }
            }

            "$TAG#dispose" -> {
                plugin.onDispose(hashCode())
                texture = null
                session = null
                scope.cancel()
                result.success(null)
            }
        }
    }

    override fun onListen(
        arguments: Any?, events: EventChannel.EventSink?
    ) {
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        eventSink = null
    }
}
